Expose an out-of-place _reduce from ProcessGroupNCCL

aashaka opened the pull request:
Summary: This change exposes an out-of-place `_reduce` API from the ProcessGroupNCCL. It allows reducing an input tensor and placing the output in a separate output tensor.

Test Plan: unittest added in a later commit

Differential Revision: D37508234

rohan-varma reviewed (commented):
Could you provide some context around why this new API is needed and what use cases it enables? Also, we usually would want unittests in the same commit adding the new functionality as a best practice :)

kwen2501 commented:
This PR seems to have duplicated code with the other PR: https://github.com/pytorch/pytorch/pull/82450
Is that needed?

Commit:
[c10d] Fix async error in batch_isend_irecv (#82450)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82450

`batch_isend_irecv` previously required the use of `torch.cuda.synchronize` to avoid data race conditions. This was because the ncclStreams were recorderd in the returned ncclWork object _before_ a ncclGroupEnd by the `_batch_p2p_manager` was issued. Thus, the `req.wait()` was effectively waiting on nothing, leading to the later operators working on incorrect intermediate data.

This fix:
- keeps track of ncclStreams to wait on, and records them in the work objects after the batch manager issues a ncclGroupEnd
- renames the `_batch_p2p_manager` to `_coalescing_manager` for generality
- removes the explicit check for NCCL backend inside `_batch_p2p_manager` in distributed_c10.py and moves the manager start/end to ProcessGroup.hpp, in order to transparently work with all process groups

Test Plan: Modified the unittest for `batch_isend_irecv` to check that received tensors are the same as expected tensors. Verified that the test fails before the change, and passes after the change.

Reviewed By: kwen2501

Differential Revision: D38100789

fbshipit-source-id: 045de67067b038db50b4155cc95b5dbcbb8f1ee7

Commit:
[PGNCCL] Expose an out-of-place _reduce to c10d (#80575)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/80575

This change exposes an out-of-place `_reduce` API from the ProcessGroupNCCL. It allows reducing an input tensor and placing the output in a separate output tensor.

Test Plan: Added `test__reduce_ops` in `caffe2/test/distributed/test_c10d_nccl.py` to test the newly exposed `_reduce` function in ProcessGroupNCCL.

Differential Revision: D37508234

fbshipit-source-id: 5830e409d4eadf13907bcd3f4914f386891c102e
